spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/TableWriter.scala

```diff
@@ -97,42 +97,53 @@ class TableWriter[T] private (
     }
   }
 
+  def batchRoutingKey(session: Session, routingKeyGenerator: RoutingKeyGenerator)(bs: BoundStatement): Any = {
+    writeConf.batchLevel match {
+      case BatchLevel.All => 0
+
+      case BatchLevel.ReplicaSet =>
+        if (bs.getRoutingKey == null)
+          bs.setRoutingKey(routingKeyGenerator(bs))
+        session.getCluster.getMetadata.getReplicas(keyspaceName, bs.getRoutingKey).hashCode() // hash code is enough
+
+      case BatchLevel.Partition =>
+        if (bs.getRoutingKey == null) {
+          bs.setRoutingKey(routingKeyGenerator(bs))
+        }
+        bs.getRoutingKey.duplicate()
+    }
+  }
+
+  def newRateLimiter = {
+    if (!OutputMetricsUpdater.detailedMetricsEnabled && writeConf.throughputMBPS < WriteConf.DefaultThroughputMBPS)
+      logError(
+        s"Detailed metrics not enabled but write throughput limiting requested at ${writeConf.throughputMBPS} MB/s." +
+          "Throughput limiting will not function properly. Please set spark.cassandra.output.metrics to true.")
+    new RateLimiter(writeConf.throughputMBPS * 1000000, 1000000)
```

Comment:
https://github.com/apache/cassandra/search?utf8=%E2%9C%93&q=1024+*+1024&type=Code
Lots of bibytes :)